### ISSUE TITLE
Docusaurus Bundle Size Heavily Impacted By Number of Routes

### ISSUE THREAD

**Issue by NEnochK** [facebook/docusaurus#7256]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [ ] I have tried the `npm run clear` or `yarn clear` command.
- [ ] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [ ] I have read the console error message carefully (if applicable).

### Description

**[Discussion from the Docusaurus Discord server](https://discord.com/channels/398180168688074762/867060369087922187/969108175548723230)**

tl;dr The number of routes/pages is directly impacting the bundle size of a Docusaurus build and it seems there are opportunities to further optimize this.

I unfortunately can't share my project/output here, but I am running into an issue where my site with ~4k pages is very slow on initial site load. After initial investigation, I found that the bundled main.xxxx.js file was (~10mb) and was littered with strings of the file paths for all the pages in our project (e.g. `/docs/path/to/file.md`). One optimization we made was to dump all markdown files into `/docs` and rely on slugs in order to shorten the length of the path string (e.g. `/docs/path/to/file.md` to `/docs/file.md`). This reduced it down to ~4mb but this is still too large. Ideally the number of routes/paths wouldn't be a bottleneck to the framework.

@Josh-Cena mentioned a possible optimization to split up registry.js/routesChunkNames.json into multiple JSON files and lazy load each only when on the path

### Steps to reproduce

1. Create a project with a large amount of routes/pages in the order of thousands.
2. Do a production build.
3. Look at the built main.xxxx.js under `build/assets/js` directory
4. Note the binary size (for me was a couple mb's) and that the file is littered with the file paths of the markdown files
5. Serve the build and note that it takes a long time to load the page initially due to this large main.xxxx.js

### Expected behavior

Expect the site to load quickly regardless of the number of pages

### Actual behavior

The initial load of the site is hindered by this large main.xxxx.js

### Self-service

- [ ] I'd be willing to fix this bug myself.

**Comment by Josh-Cena** [comment]
TBF, the number routes would always impact the entry point size, because the router has to know all available routes upfront. However, including the file paths in the bundle seems suspicious—maybe we can find a way to prune that.

**Comment by slorber** [comment]
yes, there are a few things we can improve to make the bundle size smaller.

we'll improve that over time

**Comment by Glup3** [comment]
any updates or new ideas? i stumbled upon this issue - i'm facing the same problem with 1000 routes and huge registry.js file (1 json file for each route)

**Comment by richard-lopes-ifood** [comment]
Lazy loading the pages isn't an option?

**Comment by Josh-Cena** [comment]
@richard-lopes-ifood Depends on what you mean by "lazy load". The router has to know certain metadata about the routes available to be able to load them at all, but all substantial logic of each page (its React components, for example) is only loaded when you are about to navigate to it.

**Comment by fgatti675** [comment]
Thanks to this issue I discovered why my performance was so bad.
I autogenerate 2k pages for our API, and have 3 versions.
So 6k+ routes. It was not even building anymore :(

**Comment by aeneasr** [comment]
This definitely is still a problem with current versions of Docusaurus, unfortunately.

**Comment by slorber** [comment]
@aeneasr can you help us understand which particular data is heavy on your site, and share the url?

**Comment by aeneasr** [comment]
Thank you for the answer, my message was a bit short and snarky, sorry about that. Here's the context:

We run docusaurus at https://www.ory.com/docs/welcome

The main bundle https://www.ory.com/docs/assets/js/main.41b6baa5.js (will change after rebuild) is 350kb and a big amount of content is:

```
            a378b542: [ () => o.e(49595).then(o.bind(o, 69890)), "@site/docs/kratos/concepts/security.mdx", 69890],
            a42a7080: [ () => o.e(12198).then(o.bind(o, 98009)), "@site/docs/kratos/cli/kratos-jsonnet-format.md", 98009],
```

and

```
            path: "/docs/",
            component: p("/docs/", "601"),
            routes: [{
                path: "/docs/",
                component: p("/docs/", "8db"),
                routes: [{
                    path: "/docs/",
                    component: p("/docs/", "c63"),
                    routes: [{
                        path: "/docs/account-experience",
                        component: p("/docs/account-experience", "cb9"),
                        exact: !0,
                        sidebar: "kratos"
                    }, {
                        path: "/docs/actions/integrations/hubspot",
                        component: p("/docs/actions/integrations/hubspot", "556"),
                        exact: !0,
                        sidebar: "kratos"
                    }
```

We plan on creating more pages to separate oss/saas content from one another, and it will probably bloat the bundle size even more.

The lighthouse report for the page is around 60-70 for performance, with main problems being unused JS (I assume its the routes) and unused CSS:

<img width="744" height="618" alt="Image" src="https://github.com/user-attachments/assets/543e094d-1a53-4269-822a-de40e62122bd" />

I've tried (granted with claude) to look at how this problem could be fixed by e.g. lazy loading the routes, but I know too little to be able to really build something that works and doesn't break stuff.

**Comment by slorber** [comment]
I see thanks

For v4, upgrading to React Router v7+ is one of our top priorities. Once we have upgraded, we'll be able to explore code splitting the routes into multiple slices (for example, one route slice per plugin instance, eventually split again per docs version). We need to upgrade first, but it's an optimization I have in mind.

https://remix.run/blog/fog-of-war

**Comment by aeneasr** [comment]
> I see thanks
> 
> For v4, upgrading to React Router v7+ is one of our top priorities. Once we have upgraded, we'll be able to explore code splitting the routes into multiple slices (for example, one route slice per plugin instance, eventually split again per docs version). We need to upgrade first, but it's an optimization I have in mind.
> 
> https://remix.run/blog/fog-of-war

That's huge! Looking very much forward to it  :)

**Comment by aeneasr** [comment]
While we're on the "make a feature request for this open source project" train - it would also be amazing if CSS can be extracted into critical (above the fold) and non-critical, where critical CSS is embedded on the page and the non-critical loaded asynchronous. This would also speed up page delivery by a good amount!

**Comment by yourkarma6788** [comment]
WHAT IN THE WORLD ... I CAN TELL YOU THIS MUCH, I DID NOT AUTHORIZE THIS ANS NOR DID I GIVE CONSENT FOR THIS BLOCK REQUEST

**Comment by yourkarma6788** [comment]
Raymond McKenna is illegally requesting blocklists, whitelists, and restrictions that I did not authorize. I have NEVER authorize any APIs before

**Comment by yourkarma6788** [comment]
YOU NEED TO STOP VIOLATING MY RIGHTS. I DID NOT CONSENT TO
FACEBOOK/DOCUSAURUS.
